runtimes/pythonrt/dockerclient.go

```diff
@@ -0,0 +1,298 @@
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/go-connections/nat"
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapio"
+)
+
+const (
+	runnersLabel       = "io.autokitteh.cloud.runner"
+	networkName        = "autokitteh_runners"
+	internalRunnerPort = "9293/tcp"
+)
+
+type dockerClient struct {
+	client          *client.Client
+	activeRunnerIDs map[string]struct{}
+	allRunnerIDs    map[string]struct{}
+	mu              *sync.Mutex
+	runnerLabels    map[string]string
+	logBuildProcess bool
+	logRunner       bool
+	logger          *zap.Logger
+}
+
+func NewDockerClient(logger *zap.Logger, logRunner, logBuildProcess bool) (*dockerClient, error) {
+	apiClient, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return nil, err
+	}
+
+	dc := &dockerClient{
+		client:          apiClient,
+		mu:              new(sync.Mutex),
+		runnerLabels:    map[string]string{runnersLabel: ""},
+		activeRunnerIDs: map[string]struct{}{},
+		allRunnerIDs:    map[string]struct{}{},
+		logger:          logger,
+		logBuildProcess: logBuildProcess,
+		logRunner:       logRunner,
+	}
+
+	if err := dc.SyncCurrentState(); err != nil {
+		return nil, err
+	}
+
+	return dc, nil
+}
+
+// func (d *dockerClient) close() error {
+// 	return d.client.Close()
+// }
+
+func (d *dockerClient) ensureNetwork() (string, error) {
+	inspectResult, err := d.client.NetworkInspect(context.Background(), networkName, network.InspectOptions{})
+	if err != nil {
+		if !client.IsErrNotFound(err) {
+			return "", err
+		}
+
+		noICCNetworkOptions := map[string]string{"com.docker.network.bridge.enable_icc": "false"}
+		n, err := d.client.NetworkCreate(context.Background(), networkName, network.CreateOptions{Options: noICCNetworkOptions})
+		if err != nil {
+			return "", err
+		}
+		return n.ID, nil
+	}
+
+	noICCOption := inspectResult.Options["com.docker.network.bridge.enable_icc"]
+	if noICCOption != "false" {
+		return "", errors.New("network with invalid icc, need to recreate")
+	}
+
+	return inspectResult.ID, nil
+}
+
+func (d *dockerClient) StartRunner(runnerImage string, sessionID sdktypes.SessionID, cmd []string) (string, string, error) {
+	resp, err := d.client.ContainerCreate(context.Background(),
+		&container.Config{
+			Image: runnerImage,
+			Cmd:   cmd,
+			Tty:   false,
+
+			ExposedPorts: map[nat.Port]struct{}{
+				nat.Port(internalRunnerPort): {},
+			},
+			Labels: d.runnerLabels,
+		},
+		&container.HostConfig{
+			NetworkMode:  container.NetworkMode(networkName),
+			PortBindings: nat.PortMap{internalRunnerPort: []nat.PortBinding{{HostIP: "0.0.0.0"}}},
+		}, nil, nil, "")
+
+	if err != nil {
+		return "", "", err
+	}
+
+	if err := d.client.ContainerStart(context.Background(), resp.ID, container.StartOptions{}); err != nil {
+		return "", "", err
+	}
+
+	d.mu.Lock()
+	d.allRunnerIDs[resp.ID] = struct{}{}
+	d.mu.Unlock()
+
+	go func() {
```

Comment:
Move it to a named function outside? It's big enough ...

Reply:
done